### PR TITLE
Integrate transaction coordinator to ballerina

### LINKED ISSUE #4384: Integrate micro transactor coordinator to ballerina and register initiator.
**Description:**
The microtransaction coordinator is written in ballerina. We need to integrate it into ballerina client connector execution and add initiator registration process. Coordinator service needs to be up and running which called by participants to get registered with.

### PR DESCRIPTION
**Pull request by chamil321**
PR adds micro transaction coordinator to ballerina HTTP client process. 
- if any outbound call exists within a transaction block, that transaction registers with coordinator service(internally deployed). 
- Add X-XID and X-REGISTER-AT-URL headers to the outbound response.
- Commit and abort transaction actions
- PR fixes #4384 

High-level design

microtransactions 1

**MicroTransactionCoordinator** - This class represents the coordinator internal endpoint and it holds the program file of coordinator package. An API is exposed to ballerina to process the transaction. API contains commands like commitTransaction, abortTransaction,..... etc.

**TransportTransactionManager** - This class is an interface located in ballerina-core module and it manages transport related transactions in ballerina.

**HttpTransactionManager** -  This class is implemented by the TransportTransactionManager interface and does the Http transport level operations such as commit, abort,..etc and holds the microTransaction context which returned from the coordinator. BallerinaTransactionManger is created per request for each transaction. Hence it holds the HttpTransactionManager object.

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/BLangProgramRunner.java [hunk 1]
```diff
@@ -93,6 +93,33 @@ public static void runService(ProgramFile programFile) {
...
+        for (ServiceInfo serviceInfo : servicesPackage.getServiceInfoEntries()) {
+            bContext.setServiceInfo(serviceInfo);
+            BLangFunctions.invokeFunction(programFile, serviceInfo.getInitFunctionInfo(), bContext);
+            if (bContext.getError() != null) {
+                String stackTraceStr = BLangVMErrors.getPrintableStackTrace(bContext.getError());
+                throw new BLangRuntimeException("error: " + stackTraceStr);
+            }
+            programFile.getServerConnectorRegistry().registerService(serviceInfo);
+            serviceCount++;
+        }
```
Comment (anupama-pathirage) [inline modules/ballerina-core/src/main/java/org/ballerinalang/BLangProgramRunner.java]: Shall we move this duplicated code block into a private method?

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/transactions/coordinator/2pc-commons.bal [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        Protocol[] protocols = participant.participantProtocols;
+        if (protocols != null) {
+            int j = 0;
+            while (j < lengthof protocols) {
+                Protocol proto = protocols[j];
+                var status, e = notifyParticipant(transactionId, proto.url, "abort");
+                if (e != null) {
+                    err = {msg:"Hazard-Outcome"};
+                    return;
+                } else if(status == "committed") {
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/transactions/coordinator/2pc-commons.bal]: Fix the formatting issues in .bal files. (Space after if etc) Check all the places.

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/transactions/coordinator/2pc-commons.bal [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    successful = true;
+    int i = 0;
+    while (i < lengthof participantURLs) {
+        ParticipantClient participantClient = create ParticipantClient();
+        bind participantClient with participantEP;
+
+        log:printInfo("Preparing participant: " + participantURLs[i]);
+        // If a participant voted NO then abort
+        var status, e = participantEP.prepare(transactionId, participantURLs[i]);
+        if (e != null || status == "aborted") {
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/transactions/coordinator/2pc-commons.bal]: Shall we define enum for status?

**Comment by chamil321** [comment]
The change is no longer required. Hence close the PR

**Pull request closed without merging**